examples/dnn_semantic_segmentation_ex.h

```diff
@@ -145,55 +145,129 @@ template <int N, typename SUBNET> using ares_up   = dlib::relu<residual_up<block
-template <typename SUBNET> using level4 = dlib::repeat<2,res64,res<64,SUBNET>>;
-
-template <typename SUBNET> using alevel1 = dlib::repeat<2,ares512,ares_down<512,SUBNET>>;
-template <typename SUBNET> using alevel2 = dlib::repeat<2,ares256,ares_down<256,SUBNET>>;
-template <typename SUBNET> using alevel3 = dlib::repeat<2,ares128,ares_down<128,SUBNET>>;
-template <typename SUBNET> using alevel4 = dlib::repeat<2,ares64,ares<64,SUBNET>>;
+template <
+    template<typename> class TAG1,
+    template<typename> class TAG2,
+    typename SUBNET
+>
+using resize_and_concat = dlib::add_layer<
+                          dlib::concat_<TAG1,TAG2>,
+                          TAG2<dlib::resize_to_prev<TAG1,SUBNET>>>;
+
+template <typename SUBNET> using utag0 = dlib::add_tag_layer<2100+0,SUBNET>;
+template <typename SUBNET> using utag1 = dlib::add_tag_layer<2100+1,SUBNET>;
+template <typename SUBNET> using utag2 = dlib::add_tag_layer<2100+2,SUBNET>;
+template <typename SUBNET> using utag3 = dlib::add_tag_layer<2100+3,SUBNET>;
+template <typename SUBNET> using utag4 = dlib::add_tag_layer<2100+4,SUBNET>;
+
+template <typename SUBNET> using utag0_ = dlib::add_tag_layer<2110+0,SUBNET>;
+template <typename SUBNET> using utag1_ = dlib::add_tag_layer<2110+1,SUBNET>;
+template <typename SUBNET> using utag2_ = dlib::add_tag_layer<2110+2,SUBNET>;
+template <typename SUBNET> using utag3_ = dlib::add_tag_layer<2110+3,SUBNET>;
+template <typename SUBNET> using utag4_ = dlib::add_tag_layer<2110+4,SUBNET>;
+
+template <typename SUBNET> using concat_utag0 = resize_and_concat<utag0,utag0_,SUBNET>;
+template <typename SUBNET> using concat_utag1 = resize_and_concat<utag1,utag1_,SUBNET>;
+template <typename SUBNET> using concat_utag2 = resize_and_concat<utag2,utag2_,SUBNET>;
+template <typename SUBNET> using concat_utag3 = resize_and_concat<utag3,utag3_,SUBNET>;
+template <typename SUBNET> using concat_utag4 = resize_and_concat<utag4,utag4_,SUBNET>;
 
-template <typename SUBNET> using level1t = dlib::repeat<2,res512,res_up<512,SUBNET>>;
-template <typename SUBNET> using level2t = dlib::repeat<2,res256,res_up<256,SUBNET>>;
-template <typename SUBNET> using level3t = dlib::repeat<2,res128,res_up<128,SUBNET>>;
-template <typename SUBNET> using level4t = dlib::repeat<2,res64,res_up<64,SUBNET>>;
+// ----------------------------------------------------------------------------------------
 
-template <typename SUBNET> using alevel1t = dlib::repeat<2,ares512,ares_up<512,SUBNET>>;
-template <typename SUBNET> using alevel2t = dlib::repeat<2,ares256,ares_up<256,SUBNET>>;
-template <typename SUBNET> using alevel3t = dlib::repeat<2,ares128,ares_up<128,SUBNET>>;
-template <typename SUBNET> using alevel4t = dlib::repeat<2,ares64,ares_up<64,SUBNET>>;
+// Microsoft Visual C++ compilers choke when trying to build networks that are too deep.
+// So let's limit the depth a little. On the PASCAL VOC 2012 data, the accuracy hit is
+// not even very significant.
```

Comment:
I could not quite decide if I should keep the example small enough that MSVC can build it, or have it be closer to the full U-net implementation with all the skip connections.

So at least for now, I decided to have both (using `#ifdef`s), even if the resulting code is a bit of a mess. But for simplicity, am willing to go for just one implementation.

(After this has been decided, I'll make the trained `v2` model(s) available.)

Note that on the PASCAL VOC 2012 data, there's not too much difference in terms of accuracy.